nvflare/fuel/f3/drivers/aio_http_driver.py

```diff
@@ -0,0 +1,184 @@
+# Copyright (c) 2021-2023, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+import asyncio
+import logging
+from typing import Any, Dict, List
+
+import websockets
+from websockets.exceptions import ConnectionClosedOK
+
+from nvflare.fuel.f3.comm_error import CommError
+from nvflare.fuel.f3.connection import BytesAlike, Connection
+from nvflare.fuel.f3.drivers import net_utils
+from nvflare.fuel.f3.drivers.aio_context import AioContext
+from nvflare.fuel.f3.drivers.base_driver import BaseDriver
+from nvflare.fuel.f3.drivers.driver import ConnectorInfo
+from nvflare.fuel.f3.drivers.driver_params import DriverCap, DriverParams
+from nvflare.fuel.f3.drivers.net_utils import get_tcp_urls
+from nvflare.fuel.f3.sfm.conn_manager import Mode
+from nvflare.fuel.hci.security import get_certificate_common_name
+
+log = logging.getLogger(__name__)
+
+THREAD_POOL_SIZE = 8
+MAX_MSG_SIZE = 2 * 1024 * 1024 * 1024  # 2GB
+
+
+class WsConnection(Connection):
+    def __init__(self, websocket: Any, aio_context: AioContext, connector: ConnectorInfo, secure: bool):
+        super().__init__(connector)
+        self.websocket = websocket
+        self.aio_context = aio_context
+        self.closing = False
+        self.secure = secure
+        self.conn_props = self._get_socket_properties()
+
+    def get_conn_properties(self) -> dict:
+        return self.conn_props
+
+    def close(self):
+        self.closing = True
+        self.aio_context.run_coro(self.websocket.close())
+
+    def send_frame(self, frame: BytesAlike):
+        self.aio_context.run_coro(self._async_send_frame(frame))
```

Comment:
we don't need to call `result` for this one?
When should we call result when should we not call?

Reply:
This one can't call result(). It causes deadlock of SFM handshake. This is kind of a bug in SFM. We need to fix it so non-block send_frame is not required.